src/pydvl/valuation/methods/knn_shapley.py

```diff
@@ -13,3 +13,143 @@
+    """Computes exact Shapley values for a KNN classifier.
+
+    This implements the method described in
+    (Jia, R. et al., 2019)<sup><a href="#jia_efficient_2019a">1</a></sup>.
+    It exploits the local structure of K-Nearest Neighbours to reduce the number
+    of calls to the utility function to a constant number per index, thus
+    reducing computation time to $O(n)$.
+
+    Args:
+        utility: KNNUtility with a KNN model to extract parameters from. The object
+            will not be modified nor used other than to call [get_params()](
+            <https://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html#sklearn.base.BaseEstimator.get_params>)
+        progress: Whether to display a progress bar.
+
+    """
+
+    def __init__(self, utility: KNNClassifierUtility, progress: bool = True):
+        self.utility = utility
+        self.progress = progress
+
+        config = self.utility.model.get_params()
+        self.n_neighbors = config["n_neighbors"]
+
+        del config["weights"]
+        self.helper_model = NearestNeighbors(**config)
+
+    def fit(self, data: Dataset) -> Self:
+        """Calculate exact shapley values for a KNN model on a dataset.
+
+        This fit method bypasses direct evaluations of the utility function and
+        calculates the Shapley values directly.
+
+        In contrast to other data valuation models, the runtime increases linearly
+        with the size of the test dataset.
+
+        Calculating the least core valuation is a computationally expensive task that
+        can be parallelized. To do so, call the `fit()` method inside a
+        `joblib.parallel_config` context manager as follows:
+
+        ```python
+        from joblib import parallel_config
+
+        with parallel_config(n_jobs=4):
+            valuation.fit(data)
+        ```
+
+        """
+        self.helper_model = self.helper_model.fit(data.x)
+        n_obs = len(data.x)
+        n_test = len(self.utility.test_data)
+
+        generator = zip(self.utility.test_data.x, self.utility.test_data.y)
+
+        generator_with_progress = tqdm(
+            generator,
+            total=n_test,
+            disable=not self.progress,
+            position=0,
+        )
+
+        with Parallel(return_as="generator") as parallel:
+            results = parallel(
+                delayed(_compute_values_for_one_test_point)(
+                    self.helper_model, x, y, data.y
+                )
+                for x, y in generator_with_progress
+            )
+            values = np.zeros(n_obs)
+            for res in results:
+                values += res
+            values /= n_test
+
+        res = ValuationResult(
+            algorithm="knn_shapley",
+            status=Status.Converged,
+            values=values,
+            data_names=data.data_names,
+        )
+
+        self.result = res
+        return self
+
+
+def _compute_values_for_one_test_point(
```

Comment:
As in the previous PR, I would bundle this with the class in the form of a static method.

Reply:
Done, thanks!